code/datums/components.dm

```diff
@@ -0,0 +1,92 @@
+/datum/component
+	var/enabled = TRUE      					// Enables or disables the components
+	var/dupe_mode = COMPONENT_DUPE_HIGHLANDER	// How components of the same type are handled in the same parent
+	var/list/signal_procs						// list of signals -> callbacks
+	var/datum/parent							// parent datum
+
+/datum/component/New(datum/P, ...)
+	var/dm = dupe_mode
+	if(dm != COMPONENT_DUPE_ALLOWED)
+		var/datum/component/old = P.GetExactComponent(type)
+		if(old)
+			switch(dm)
+				if(COMPONENT_DUPE_HIGHLANDER)
+					P.RemoveComponent(old)
+					old = null	//in case SendSignal() blocks
+				if(COMPONENT_DUPE_UNIQUE)
+					qdel(src)
+					return
+	LAZYADD(P.datum_components, src)
+	SendSignal(COMSIG_COMPONENT_ADDED, list(src), FALSE)

```

Comment:
Ironically I only said it because I misunderstood and thought the second argument was the receivers.

I'll claim it was my subconscious thought that bubbled the idea up and my conscious thought just misinterpreted it.